designs/2023-test-rule-errors/README.md

```diff
@@ -0,0 +1,313 @@
+
+It can be tricky to get schemas to perfectly represent what the allowed input should be, especially when no automated testing exists around this today. Schemas are often too lenient, allowing extra input that the rule implementation ignores or doesn't handle properly. This can result in rules silently misbehaving, not respecting the wishes or intentions of the user's configuration, or just crashing.
+
+By enabling testing of schemas, we'll make it easier to write higher-quality schemas and rules, thus improving the user experience.
+
+## Detailed Design
+
+<!--
+   This is the bulk of the RFC.
+
+   Explain the design with enough detail that someone familiar with ESLint
+   can implement it by reading this document. Please get into specifics
+   of your approach, corner cases, and examples of how the change will be
+   used. Be sure to define any new terms in this section.
+-->
+
+Add support to the rule tester for a new `fatal` test case array in addition to `valid` and `invalid`.
+
+Example with a schema validation error because a non-existent option is passed:
+
+```json
+{
+    "fatal": [
+        {
+            "code": "foo();",
+            "options": [{ "checkFoo": true, "nonExistentOption": true }],
+            "error": {
+                "message": "Value {\"checkFoo\":true,\"nonExistentOption\":true} should NOT have additional properties.",
+                "name": "SchemaValidationError"
+            }
+        }
+    ]
+}
+```
+
+To test a custom exception thrown by a rule, simply use the exception text as the message: `"message": "Some custom exception message from the rule."` and/or the name of the exception class as the `name`.
+
+- The `error` object groups the properties from the exception together nicely.
+- Similar to `message` and `messageId` in `invalid` test cases, only one of `message` and `name` is required. In particular, if the user doesn't care to test the full exception `message` text, the exception class `name` can be used instead.
+- `message` can be provided as a string or a regular expression (regexp), while `name` can only be provided as a string.
+- `message` comes from `err.message`
+- `name` comes from `err.name` or `err.constructor.name` in case either has a custom value (besides just `Error`)
+
+Examples of how exceptions would be thrown with different exception classes that could be distinguished by `name`:
+
+- `throw new SchemaValidationError('...should NOT have additional properties.')` (custom error type thrown by schema validation)
+- `throw new SomeCustomRuleError('...')` (custom error type thrown by the rule implementation)
+- `throw new Error('...')` (basic error thrown by the rule implementation)
+
+Differences between `fatal` test cases and `invalid` test cases:
+
+- `code` is not required (the code may be irrelevant when testing options)
```

Comment:
In my draft PR (https://github.com/eslint/eslint/pull/16823/files#diff-1d5e36429b1a08e3baa8311c3c702f2edf5e3c5e542771dcd6a6bef4c43727f5R1173), I solved this issue by using a placeholder `(Test Case #x)` when there's no `name` nor `code`.

```js
fatal.name || fatal.code || `(Test Case #${index + 1})`
```

I prefer this over requiring the user to come up with a `name` anytime they want to omit `code`, which feels like it would be a burden.

Is that okay?

Reply:
This solves the test case name (first argument of `it`). I was thinking about the `code` we are passing to [`linter.verify`](https://github.com/eslint/eslint/blob/67865a064cc1a4e320030299edc1cfdd1f9ac3b8/lib/rule-tester/rule-tester.js#L690).

Reply:
I think `code` can be an empty string.

Reply:
> I think `code` can be an empty string.

If `code` property isn't specified, we'll pass an empty string to `linter.verify`? Sounds good to me.

Reply:
Yup, exactly.

Reply:
Will address remaining issues in the next week.

Reply:
I added a note about using an empty string for `code`.